src/structures/paging/mapper/recursive_page_table.rs

```diff
@@ -47,6 +47,11 @@ impl<'a> RecursivePageTable<'a> {
     /// - The page table must be active, i.e. the CR3 register must contain its physical address.
     ///
     /// Otherwise `Err(())` is returned.
+    ///
+    /// ## Safety
+    ///
+    /// Creating a recursive page table with recursive index 511 is unsound
+    /// because calculating the end ptr of the structure causes an overflow.
```

Comment:
I'm not sure about the wording here. It should not be unsound to create the `RecursivePageTable` struct. If anything, it is unsound to create the `table` reference that is passed in. However, I'm not even sure about that.

The [chapter about undefined behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) in the Rust reference does not mention any requirement that the end pointer of a structure must be valid. So the unsoundness might be in the standard library. For example, the [`iter_mut` implemenation for slices](https://github.com/rust-lang/rust/blob/314c39d2ea07d8b50649149358ebeb1a6bd09179/library/core/src/slice/iter.rs#L211-L232) calls the unsafe `ptr.add(slice.len()`. The [`pointer::add` method](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.add) cleary requires that there must be no overflow:

> The offset being in bounds cannot rely on “wrapping around” the address space. That is, the infinite-precision sum must fit in a usize.

Given that the slice iterator methods are safe, I think they should check for this requirement instead of triggering undefined behavior. Alternatively, they could use `wrapping_add` instead.

Either way, I think we should discuss this with the opsem team. I'll create a thread on Zulip.

Reply:
I started a Zulip thread at https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/end-of-address-space/near/366416265.

Reply:
The opsem team confirmed that allocating the last byte of the address space is considered unsound.

I tried to update the comment to make the root cause of the UB a bit clearer. I also think that we should link the Zulip and GitHub threads. What do you think?

```suggestion
    /// Note that creating a `PageTable` with recursive index 511 is unsound
    /// because allocating the last byte of the address space can lead to pointer
    /// overflows and undefined behavior. For more details, see the discussions
    /// [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/end-of-address-space)
    /// and [in the `unsafe-code-guidelines ` repo]https://github.com/rust-lang/unsafe-code-guidelines/issues/420).
```

Reply:
Looks good, thanks!